File: worker/src/main/java/com/github/k0kubun/gitstar_ranking/repository/PaginatedOrganizations.java
package com.github.k0kubun.gitstar_ranking.repository;

import com.github.k0kubun.gitstar_ranking.model.Organization;
import com.github.k0kubun.gitstar_ranking.repository.dao.OrganizationDao;

import java.util.List;

import org.skife.jdbi.v2.Handle;

// This class does cursor-based-pagination for organizations order by stargazers_count DESC.
public class PaginatedOrganizations
{
    private static final int PAGE_SIZE = 5000;

    private final OrganizationDao orgDao;
    private Integer lastMinStars;
    private Integer lastMinId;

    public PaginatedOrganizations(Handle handle)
    {
        orgDao = handle.attach(OrganizationDao.class);
        lastMinStars = null;
        lastMinId = null;
    }

    public List<Organization> nextOrgs()
    {
        List<Organization> orgs;
        if (lastMinId == null && lastMinStars == null) {
            orgs = orgDao.starsDescFirstOrgs(PAGE_SIZE);
        }
        else {
            orgs = orgDao.starsDescOrgsAfter(lastMinStars, lastMinId, PAGE_SIZE);
        }
        if (orgs.isEmpty()) {
            return orgs;
        }

        Organization lastOrg = orgs.get(orgs.size() - 1);
        lastMinStars = lastOrg.getStargazersCount();
        lastMinId = lastOrg.getId();
        return orgs;
    }
}

File: worker/src/main/kotlin/com/github/k0kubun/gitstar_ranking/db/PaginatedOrganizations.kt
package com.github.k0kubun.gitstar_ranking.db

import com.github.k0kubun.gitstar_ranking.core.Organization
import com.github.k0kubun.gitstar_ranking.core.StarsCursor
import org.jooq.DSLContext

private const val PAGE_SIZE = 5000

// This class does cursor-based-pagination for organizations order by stargazers_count DESC.
class PaginatedOrganizations(private val database: DSLContext) {
    private var lastMinStars: Long? = null
    private var lastMinId: Long? = null

    fun nextOrgs(): List<Organization> {
        val orgs: List<Organization> = if (lastMinId != null && lastMinStars != null) {
            OrganizationQuery(database).orderByStarsDesc(
                limit = PAGE_SIZE,
                after = StarsCursor(id = lastMinId!!, stars = lastMinStars!!)
            )
        } else {
            OrganizationQuery(database).orderByStarsDesc(limit = PAGE_SIZE)
        }
        if (orgs.isEmpty()) {
            return orgs
        }
        val lastOrg = orgs[orgs.size - 1]
        lastMinStars = lastOrg.stargazersCount
        lastMinId = lastOrg.id
        return orgs
    }
}
